### PR TITLE
dns_resolver: add stats for c-ares based DNS resolver

### PR DESCRIPTION
**Pull request by surki**
Commit Message: dns_resolver: add stats for c-ares based DNS resolver
Additional Description: Recently we ran into an outage caused by DNS lookup failures. This was caused by a combination of `k8s/kube-proxy + iptables + conntrack` bug (which blackholed an opened UDP socket until envoy restart). Due to lack of metrics on DNS lookups, alerting was not feasible. This adds metrics support for c-ares based DNS lookups, similar to apple based DNS resolver.
Risk Level: Low
Testing: Manual, Unit tests
Docs Changes: Done
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by surki** [comment]
@phlax wondering if you can re-trigger the pipeline, they are stuck (unrelated to this change), repokitteh is not able to re-run failed jobs

/retest

File: test/extensions/network/dns_resolver/cares/dns_impl_test.cc [hunk 1]
```diff
@@ -1340,26 +1448,42 @@ TEST_P(DnsImplTest, PendingTimerEnable) {
...
   Event::MockTimer* timer = new NiceMock<Event::MockTimer>();
   EXPECT_CALL(dispatcher, createTimer_(_)).WillOnce(Return(timer));
-  resolver_ = std::make_shared<DnsResolverImpl>(config, dispatcher, vec);
+  resolver_ = std::make_shared<DnsResolverImpl>(config, dispatcher, vec, stats_store_);
   Event::FileEvent* file_event = new NiceMock<Event::MockFileEvent>();
   EXPECT_CALL(dispatcher, createFileEvent_(_, _, _, _)).WillOnce(Return(file_event));
   EXPECT_CALL(*timer, enableTimer(_, _));
   EXPECT_NE(nullptr, resolveWithUnreferencedParameters("some.bad.domain.invalid",
                                                        DnsLookupFamily::V4Only, true));
+  checkStats(0 /*resolve_total*/, 1 /*pending_resolutions*/, 0 /*not_found*/,
```
Comment (htuch) [inline test/extensions/network/dns_resolver/cares/dns_impl_test.cc:1457]: Probably worth adding a test to track >1 pending resolution and satisfactory popping of the resolutoins back to zero.

**Review by htuch — CHANGES_REQUESTED** [review]
Seems useful, thanks for  the contribution.
/wait

File: docs/root/intro/arch_overview/upstream/dns_resolution.rst [hunk 2]
```diff
@@ -23,6 +23,18 @@ Envoy provides DNS resolution through extensions, and contains 3 built-in extens
 
 For an example of a built-in DNS typed configuration see the :ref:`HTTP filter configuration documentation <config_http_filters_dynamic_forward_proxy>`.
 
+The c-ares based DNS Resolver emits the following stats rooted in the ``dns.cares`` stats tree:
+
+  .. csv-table::
+    :header: Name, Type, Description
+    :widths: 1, 1, 2
+
+    resolve_total, Count, Number of DNS queries
```
[the remark this answers is not in the document]
Comment (surki) [reply docs/root/intro/arch_overview/upstream/dns_resolution.rst:32]: do you want me to add that? I could.

[the remark this answers is not in the document]
Comment (htuch) [reply docs/root/intro/arch_overview/upstream/dns_resolution.rst:32]: Maybe good as a followup, doesn't need to be in this PR. Will defer to other owners of the Apple resolver, @yanavlasov @mattklein123 on whether some of these stats should be ported in a followup.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**